Temporarily revert star annotation in call signatures - remove all internal xarray imports

Zeitsperre opened the pull request:
<!--Please ensure the PR fulfills the following requirements! -->
<!-- If this is your first PR, make sure to add your details to the AUTHORS.rst! -->
### Pull Request Checklist:
- [x] This PR addresses an already opened issue (for bug fixes / features)
    - This PR fixes #2120
    - This PR fixes #2113 
- [x] Tests for the changes have been added (for bug fixes / features)
  - [ ] (If applicable) Documentation has been added / updated (for bug fixes / features)
- [x] CHANGELOG.rst has been updated (with summary of main changes)
  - [x] Link to issue (:issue:`number`) and pull request (:pull:`number`) has been added

### What kind of change does this PR introduce?

* Reverts star-annotated call signatures (not supported for Python 3.10)
*`xclim` now uses directly `operator` instead of using `xarray`'s derived `get_op` function. A refactoring in `xarray` had changed the position of `get_op` which caused a bug.
    + All other uses of `xarray`'s internal API were also removed (:pull:`2116`).
* Updates the pre-commit hooks.

### Does this PR introduce a breaking change?

No.

Zeitsperre commented:
This PR needs the changes from #2114

aulemahal commented:
@Zeitsperre My last commit removed all xarray imports that were not from their "public" API, hoping that avoids all further problemes like we add with xarray 2025.3.

To do so, I had to copy a few lines from xarray for the `merge_attrs_drop_conflicts` and `get_temp_dimname` functions. @coxipi  you might want to the the same as I did for xsdba, even though this is not a bug right now, just futurproofing.

I also had to remove `DataArrayResample` and `DatasetResample` from the signature of `xc.core.calendar.time_bnds`. They are still mentionned in the docstring... Not the best, but I don't see how I can have them in the signature without importing them at runtime....

Zeitsperre commented:
> I also fixed an issue with setting options that are deep dictionaries. And fixed the issue with the missing methods. Flox is now enabled by default for resampling with `first` and `last`, but the way xarray calls flox is broken for datetime data in some cases. I simply used another resampling method as the line in `expected_count` only cares about the output coordinates. (This call is done because the alternative was to dig into private xarray objects)

Thanks for the fixes! Think you could mention this in the Changelog as a (potentially) breaking change?

aulemahal commented:
> as a (potentially) breaking change?

I don't think the user is impacted at all! This change does not change anything in the function's outputs.

I found more xarray internals and removed them. I had to reimplement a version of  `days_in_year` to do so, but the version I copied from xarray is pretty short.

The last xarray internal calls I see, but couldn't/wouldn't remove are:
- `xr.core.common._contains_cftime_datetimes` and `xr.conventions.encode_cf_variable` in `xclim.sdba.map_blocks` : I think the whole block of code can be removed, but I will do so in a PR to `xsdba` instead.
- `xr.coding.cftime_offsets.to_offset` in a few places : we need the returned `Offset` object, which itself isn't public API anyway. No way I am reimplementing this whole machinery in xclim.

Finally, I also had to remove the annotations from `xclim.sdba.Grouper.group`. I think returning a `GroupBy` object is ok, but it isn't itself public API, only its methods are... It could be moved around in the module without warning.
